packages/kbn-custom-events/src/index.ts

```diff
@@ -0,0 +1,180 @@
+            esOrgId,
+            version: kibanaVer,
+            versionMajor: parsedVer[0] ?? -1,
+            versionMinor: parsedVer[1] ?? -1,
+            versionPatch: parsedVer[2] ?? -1,
+          })
+        );
+        this.fullStory = fullStory;
+      }
+    } catch (e) {
+      // eslint-disable-next-line no-console
+      console.error(`[custom events] Could not call initialize due to error: ${e.toString()}`, e);
+    }
+
+    return !!this.fullStory;
+  }
+
+  private getFullstoryType(value: FullstoryEventValue) {
+    // For arrays, make the decidion based on the first element
+    const v = isArray(value) ? value[0] : value;
+    switch (typeof v) {
+      case 'string':
+        return 'str';
+      case 'number':
+        return isInteger(value) ? 'int' : 'real';
+      case 'boolean':
+        return 'bool';
+      case 'object':
+        if (v instanceof Date) {
+          return 'date';
+        }
+    }
+  }
+
+  private getCurrentMemoryState() {
+    // Get performance information from the browser (non standard property
+    // @ts-expect-error 2339
+    const memory = window.performance.memory;
+    let memInfo = {};
+    if (memory) {
+      memInfo = {
+        jsHeapSizeLimit: memory.jsHeapSizeLimit,
+        jsHeapSizeTotal: memory.totalJSHeapSize,
+        jsHeapSizeUsed: memory.usedJSHeapSize,
+      };
+    }
+    return memInfo;
+  }
+
+  private formatContext(context: Record<string, FullstoryEventValue>) {
+    // format context keys as required for env vars, see docs: https://help.fullstory.com/hc/en-us/articles/360020623234
+    return mapKeys(context, (value, key) => {
+      const snakeCasedKey = snakeCase(key);
+      const type = this.getFullstoryType(value);
+      return [snakeCasedKey, type].join('_');
+    });
+  }
+
+  public setUserContext(context: Record<string, FullstoryEventValue>) {
+    if (!this.fullStory) return;
+
+    try {
+      this.fullStory.setUserVars(this.formatContext(context));
+    } catch (e) {
+      // eslint-disable-next-line no-console
+      console.error(
+        `[custom events] Could not report custom event due to error: ${e.toString()}`,
+        e
+      );
+    }
+  }
+
+  public setCustomEventContext(context: Record<string, FullstoryEventValue>) {
+    if (!this.fullStory) return;
+
+    Object.keys(context).forEach((key) => {
+      if (context[key] === undefined) {
+        this.clearCustomEventContext(key);
+      } else {
+        this.customEventContext[key] = context[key];
+      }
+    });
+  }
+
+  private clearCustomEventContext(fieldName: string | string[]) {
+    if (!this.fullStory) return;
+    const fields = typeof fieldName === 'string' ? [fieldName] : fieldName;
+    fields.forEach((field) => {
+      if (this.customEventContext.hasOwnProperty(field)) {
+        delete this.customEventContext[field];
+      }
+    });
+  }
+
+  reportCustomEvent = (eventName: string, params: Record<string, FullstoryEventValue> = {}) => {
```

Comment:
How are we going to retrieve data from fullstory if we a user wants to see an example of a telemetry payload?  Are we storing the data on the local machine for some time?

Reply:
In the requirements that we are setting for Event-based telemetry, we don't plan on persisting the events. Following VSCode's approach, if the users want to see them, they can enable a specific log (disabled by default).